Feature cm cloud message

mlohvynenko opened the pull request.

mlohvynenko committed:
common: cloudprotocol: add comparison operators for states

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add cloud message structures

This patch adds structures that represents messages
received and sent by the cloud.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
cm: communication: add message handler interface

This patch adds an interface that is capable of handling messages
received from by the communication module.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: types: update instance status structure

This patch aligns the instance status structure with
the cloud protocol.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add unit config status structure

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: tools: mark enum stringer class constructors as constexpr

This patch enhances the usability of the EnumStringer class
in constant expressions.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add monitoring message

This patch defines monitoring message from the cloud protocol.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: log: remove log message type enum

This patch removes redundant log message type enum
as cloud protocol message type enum contains these values.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: align request log message with protocol

This patch aligns the request log message with the cloud protocol:
- Upload options are moved out from the log filter
to the request log message;
- Log type is changed to log upload type in the upload options struct.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
cm: communication: update interface to have a single send method

This patch updates communication interface to have a single method
for sending messages that will accept variant type of cloud message.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add supported cloud message type enum values

This patch updates cloud message type enum with all
supported values.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add alerts cloud message

This patch adds alerts cloud message structure.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: crypto: define certificate chains count constant

This patch defines a constant for the maximum number
of certificate chains.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add certificates messages

This patch implements certificates cloud protocol messages.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add provisioning cloud messages

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: types: move bearer token len constant to types

This patch moves the bearer token length constant
to the types header to make it more accessible.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add service discovery messages

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add cloud protocol envvars messages

This patch adds support for cloud protocol
environment variables messages.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: types: add failed item status enum value

This patch adds failed item status enum value.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

al1img commented on include/aos/common/cloudprotocol/certificates.hpp:
```diff
@@ -0,0 +1,322 @@
+            "sm",
+            "cm",
+            "iam",
+            "host-device",
```
Not sure we use host-device, remote-device and azure-iot. Need to check with cloud team.

al1img commented on include/aos/common/cloudprotocol/certificates.hpp:
```diff
@@ -0,0 +1,322 @@
+/**
+ * Maximum number of certificates per node.
+ */
+static constexpr auto cCertificatesPerNodeCount = AOS_CONFIG_CLOUDPROTOCOL_CERTIFICATES_PER_NODE_COUNT;
```
Actually it equals to CertType enum count.

al1img commented on include/aos/common/types.hpp:
```diff
@@ -393,6 +393,11 @@ static constexpr auto cExposedPortLen = cPortLen + cProtocolNameLen;
+/**
+ * Bearer token len.
+ */
+constexpr auto cBearerTokenLen = AOS_CONFIG_CLOUDPROTOCOL_BEARER_TOKEN_LEN;
```
Where is it required also?

al1img commented on include/aos/cm/communication/communication.hpp:
```diff
@@ -16,6 +16,31 @@ namespace aos::cm::communication {
+/**
+ * Connection events subscriber interface.
+ */
+class ConnectionEventsSubscriberItf {
```
I guess we have such interface already. See common/connectionsubsc.hpp

mlohvynenko commented on include/aos/common/types.hpp:
```diff
@@ -610,6 +610,7 @@ class ItemStatusType {
         eRemoving,
         eRemoved,
         eError,
+        eFailed,
```
https://github.com/aosedge/aos_protocols/blob/3e50bfa9a9bcd038bf15bc78389949d2a8dcee46/src/cloud_common/protocols/unit/types.py#L16

To fulfill cloud protocol needs, I've added error enum (for unit config status)

mlohvynenko committed:
cm: launcher: nodemanager: increase state checksum size

sha3 224 constant holds size of a byte array, this patch
multiplies it by 2 to hold a hex representation of the checksum
in a static string object.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
build: add parallel option to specify number of jobs

This patch adds an ability to specify the number of
parallel jobs for the build process using
the --parallel option in the build.sh script.
This allows users to optimize build times based on
their system's capabilities.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: move unit status to cloudprotocol

This patch moves unit status structure to the cloudprotocol module
and updates unit status structure usages in source code.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add unit config structure

This patch adds unit config structure to the cloud protocol.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
common: cloudprotocol: add desired status message

This patch adds the desired status message to the cloud protocol.
In this patch, only the unit config field of the desired status
message is added.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

al1img reviewed (commented):
`Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>`

mykola-kobets-epam commented:
`Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>`
